[Enterprise Search] Add Connector Sync Job POST API

timgrein opened the pull request:
This PR adds the `ConnectorSyncJob` model and a corresponding `POST` endpoint for creating a new sync job.

As the current API spec expects an `id` field inside a connector sync job, which is the same value as the internal `_id` I needed to use a workaround for generating an id to avoid doing the following: Index sync job without id -> fetch this sync job again -> set the id -> index again. I've added a comment that this could be mitigated, if we remove the `id` field as it's also not specified in the [Connectors Protocol](https://github.com/elastic/connectors/blob/main/docs/CONNECTOR_PROTOCOL.md), but nothing blocking right now, just to raise the awareness. Or maybe there's a more obvious/ES way of achieving, what I want to do with this workaround?

I've also added `getSyncJobConnectorInfo` intentionally to `ConnectorSyncJobIndexService` for now as I don't want interfere with Jedr's work. I guess that's fine for now and can be addressed later, but happy to discuss that (moving that to `ConnectorIndexService` or using something Jedr probably builds down the line for another endpoint).

dnhatn committed:
Replace Block.Ref with Block reference count (#102513)

Block.Ref was introduced to help memory management for Evaluator. 
However, we can replace it with the Block referencing count now.

DaveCTurner committed:
AwaitsFix for #101095

DaveCTurner committed:
Introduce RefCounted#mustIncRef (#102515)

In several places we acquire a ref to a resource that we are certain is
not closed, so this commit adds a utility for asserting this to be the
case. This also helps a little with mocks since boolean methods like
`tryIncRef()` return `false` on mock objects by default, but void
methods like `mustIncRef()` default to being a no-op.

breskeby committed:
Check docker env by probing docker (#102558)

For now we probe ddocker functionality when using minio testcontainer.
We need to adjust our docker availability matrix.

Should mute tests and address #102532

dnhatn committed:
Update doc fields for pow and date_extract function (#102554)

These two generated files have not been updated and committed.

breskeby committed:
Fix Nullpointer in test container tests and setup logging (#102561)

related to #102532

idegtiarenko committed:
Update metric names to follow open telemetry conventions (#102529)

Update metric names to follow conventions described in https://opentelemetry.io/docs/specs/semconv/general/metrics/

heissanjay committed:
Add support for configuring proxy scheme in S3 client settings and EC2 discovery plugin (#102495)

Closes #101873

afoucret committed:
Rename the inference rescorer into learn_to_rank (#102540)

ldematte committed:
Moving NodeInfo internal version storage to String (#100746)

NodeInfo version as string. Moving from Version to Build.version(), which means also non-semantic versions. Particular care was needed for ESRestTestCase Version parsing, to make it work in a serverless environment, removing Version whenever possible and adding proper fallbacks when needed.

thecoop committed:
Move and consolidate more node construction logic (#102302)

DaveCTurner committed:
Renamings/cleanups in snapshots code (#102564)

Extracted from #101717

DaveCTurner committed:
Check that every ISRT has a primary (#102560)

breskeby committed:
Fix this-escape -Werror (#102574)

Using newer java versions expose a -Werror failure in building
MinioTestContainer

DaveCTurner committed:
Add SnapshotInProgressAllocationDecider unit tests (#102572)

This allocation decider has no unit tests today, but it would be useful
to have some so that they can be used to support changes to its logic.
This commit adds the missing test suite.

ywangd committed:
[Test mute] AwaitsFix #102547

jan-elastic committed:
Respect ChunkedDataExtractor's query in the wrapped ScrollDataExtractor. (#102492)

* Respect ChunkedDataExtractor's query in the wrapped ScrollDataExtractor.

* Update docs/changelog/102492.yaml

* Add+fix ChunkedDataExtractorTests

* Close the scroll in the preview datafeed endpoint

* Add extraFilters as constructor param of the DataExtractors

* Update docs/changelog/102492.yaml

Co-authored-by: David Roberts <dave.roberts@elastic.co>

---------

Co-authored-by: David Roberts <dave.roberts@elastic.co>

thecoop committed:
Remove DesiredNodesSettingsValidator from node construction (#102573)

It doesn't need to be injected anymore, it can just be used directly

ywangd committed:
Allow more plugin threadpool for blob operations (#102567)

Add one more threadpool from stateless plugin

Relates: #102277

jfreden committed:
Expose the invalidation_time field in Get/Query ApiKey APIs (#102472)

This PR exposes `invalidation_time` in responses from
[GetApiKey](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-get-api-key.html)
and
[QueryApiKey](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-query-api-key.html)
API To address the first half of
https://github.com/elastic/elasticsearch/issues/92404.

As a follow up to this PR `invalidation_time` needs to be supported as a
valid [query
value](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-query-api-key.html#security-api-query-api-key-request-body)
for the `QueryApiKey` API. There will also be a separate PR to add it to
the API docs for
[GetApiKey](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-get-api-key.html)
and
[QueryApiKey](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-query-api-key.html).

DaveCTurner committed:
Log exceptions thrown during chunked encoding (#102585)

Today on exception we `assert false` in the `finally` block, but that
means we don't get to see the exception that caused the problem. This
commit adds a little extra logging to help there.

michalpristas committed:
Added beat.stats.libbeat.pipeline.queue.max_events (#102570)

ldematte committed:
REST tests: GET features from currently running cluster (#102546)

edsavage committed:
[ML] Add cluster settings to override ML autoscaling (#102413)

Add the concept of an autoscaling "dummy" entity, comprised of the xpack.ml.dummy_entity_memory and xpack.ml.dummy_entity_processors settings. These are intended to be used in extreme cases to trigger autoscaling events.

martijnvg committed:
Additional minor cleanup of org.elasticsearch.search.aggregations package (#102293)

* Convert Collections.sort() to List.sort()
* Use Collections.addAll() instead of manually invoking add()
* Use Map.merge() / Map.computIfAbsent()
* Use primitive double over Double
* Replace some lambdas with method references.
* Replaces for loops with index variable to use foreach iteration style for loops.

martijnvg committed:
Unmute DownsampleActionIT#testTsdbDataStreams test (#102334)

The likely cause is that the downsample step wasn't executed, because the index.time_series.end_time index setting was after current time.

Two tweaks to avoid this failure:
* Lower index.look_ahead_time
* Increase wait time in waitAndGetRollupIndexName()

Closes #100271

leemthompo committed:
[DOCS] Add ES quickstart (#102226)

* [DOCS] TEST restore quickstart

* Use up to date Docker instructions, minor user-friendly modifications

* Use books dataset, update verbiage, add examples

* Update verbiage

* Updated Elasticsearch 'Getting Started' docs: added SSL, Docker setup, Python resources, and expanded next steps

* minor formatting

* Collapse responses, TODO comment tests

* Add request tests

* Edit superfluities

* Apply suggestions

Co-authored-by: István Zoltán Szabó <istvan.szabo@elastic.co>

* Update docs/reference/tab-widgets/quick-start-install.asciidoc

Co-authored-by: István Zoltán Szabó <istvan.szabo@elastic.co>

---------

Co-authored-by: István Zoltán Szabó <istvan.szabo@elastic.co>

thecoop committed:
Remove some more old references to Version constants (#102526)

afoucret committed:
Add params to the LTR rescorer (#102566)

dnhatn committed:
Track blocks in enrich operators (#102184)

This change tracks Blocks in enrich server operators. With this change, 
all Blocks in the enrich feature should be properly tracked with BlockFactory.

dnhatn committed:
Wrap user context when wait for async actions (#102550)

Since #102508, a Driver will wait for async actions, and its listener 
can be completed by an async action. An async action can run with a
different thread context than the current user context. This PR reserves
the current thread context when waiting for async actions.

Closes #101095
Relates #102508

dnhatn committed:
Track blocks of intermediate state of aggs (#102562)

This PR tracks blocks created when evaluating the intermediate state of aggregation.

rockdaboot committed:
[Profiling] Add new optional CO2 request params (#102576)

* [Profiling] Add new optional CO2 request params

We add new request params for the profiling API

co2_per_kwh
datacenter_pue
per_core_watt

rockdaboot committed:
[Profiling] Fix Azure region names (#102579)

* [Profiling] Fix Azure region names

Fixes the Azure region names to be all lowercase and without spaces.
This matches the region names reported by the profiling host agent.

Commit:
[Transform] Consider search context missing exceptions as recoverable (#102602)

original-brownbear committed:
Fix ref counts for MultiSearchResponse not released in tests (#102604)

Part of the effort to fix search response leaks is to fix these. Fixed
all that I could easily find in tests. Production changes incoming once
the dependencies for those are fixed.

part of #102030 but no fancy utility here like for search responses
since we don't have so many use cases an none of them are tricky.

iverase committed:
Remove explicit SearchResponse references from `org.elasticsearch.indices` and   `org.elasticsearch.indexing` (#102536)

iverase committed:
Remove explicit SearchResponse references from `org.elasticsearch.similarity` and  `org.elasticsearch.versioning` (#102542)

original-brownbear committed:
Rationalize ref-counting around ChannelActionListener (#102551)

This listener was behaving differently from other listeners in that it
would effectively result in decRef by 1 when invoked. This worked out so
far by accounting for this fact in calling code but is not maintainable
now that more and more ref-counted things are going to be passed into
it => make it behave like any other listener and be neutral in respect
to ref count by acquiring the reference that the channel will release.

Co-authored-by: Alexander Spies <alexander.spies@elastic.co>

iverase committed:
Remove explicit SearchResponse references from `org.elasticsearch.recovery` and   `org.elasticsearch.routing` (#102539)

relates https://github.com/elastic/elasticsearch/issues/102030

original-brownbear committed:
Add RestChunkedToXContentListener version that releases transport responses (#102607)

This is needed as part of #102030, we need to be able to release
search/multi-search and maybe other transport messages after they've
been serialized as REST responses.

original-brownbear committed:
Revert "Rationalize ref-counting around ChannelActionListener (#102551)" (#102609)

This reverts commit 6f72a1c83b5679b752c7cc6382ea0f838eac7e43 #102551 ,
sadly this is still causing ESQL test failures and we need to fix ESQL
some more before we can continue here.

original-brownbear committed:
Delete all unused private methods (#98111)

Pretty straight forward dead-code cleanup I think. Just delete all
private methods or methods in private classes that aren't used.

original-brownbear committed:
Cleanup dead code from SearchHit (#102616)

Using a search hit is not used any longer, also pre-7.8 serialization
is dead code now.

jfreden committed:
Add support for invalidation timestamp in QueryApiKey (#102590)

This is a follow up PR from
https://github.com/elastic/elasticsearch/pull/102472. This adds the
ability to use `invalidation` timestamp as a valid [query
value](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-query-api-key.html#security-api-query-api-key-request-body)
in the QueryApiKey API.

abdonpijpelink committed:
Update path-settings.asciidoc (#102615) (#102624)

The "the" is non-necessary and meaningless.

(cherry picked from commit 326421773f1ff60df7a116eafed4e2d53a22b8bb)

Co-authored-by: Ali <115415312+xogoodnow@users.noreply.github.com>

abdonpijpelink committed:
[DOCS] Examples for ES|QL DISSECT and WHERE (#102591)

* DISSECT examples

* WHERE examples

* Remove references to empty keys

* Fix non-deterministic test

astefan committed:
Test for #102393 (#102610)

abdonpijpelink committed:
Update intro.asciidoc (#102549) (#102626)

* Update intro.asciidoc

Added vector database use cases to the list

* Update docs/reference/intro.asciidoc

---------

Co-authored-by: Abdon Pijpelink <abdon.pijpelink@elastic.co>
(cherry picked from commit 8e78160f399463483573bdfe7cf5a7708e80851a)

Co-authored-by: addytripathi <47859340+addytripathi@users.noreply.github.com>

leemthompo committed:
[DOCS] Update get started button on landing page (#102627)

abdonpijpelink committed:
[DOCS] ES|QL: add full-text search limitation (#102150)

* [DOCS] ES|QL: add full-text search limitation

* Improve wording

DaveCTurner committed:
Assert aborted snapshots have no pending shard work (#102621)

When a snapshot is aborted, all shard snapshots must either be complete
(`SUCCESS`, `FAILED` or `MISSING`) or else they must be in state
`ABORTED` while the data node finishes its work. It seems we do not
check this invariant anywhere today, so this commit adds an assertion of
this invariant.

ldematte committed:
Revert non-semantic NodeInfo (#102636)

kingherc committed:
Add warning on desired balancer heuristics (#102633)

To avoid changing them.

javanna committed:
Enable inter-segment concurrency for terms aggs (#101390)

This commit enables inter-segment search concurrency for terms aggs, when the cardinality of the field being aggregated on is lower than the shard size. This is to avoid precision errors compared that would be caused by parallelizing the executor across slices for fields with high cardinality.

For terms agg that are ordered by key, we still take cardinality into account to parallelize only low cardinality fields and avoid performance overhead caused by parallelizing high cardinality fields.

salvatore-campagna committed:
Export circuit breaker trip count as a counter metric (#101423)

Here we export both parent and children circuit breaker trip counts as metrics so that we can collect their values using APM. We expose a counter for the trip count of the parent circuit breaker and a counter for each trip count of children circuit breakers including:

* field data circuit breakers
* per-request circuit breakers
* in-flight requests circuit breakers
* custom circuit breakers used by plugins (EQL and Machine Learning)

The circuit breaker metrics include:

* es.breaker.parent.trip.total
* es.breaker.field_data.trip.total
* es.breaker.request.trip.total
* es.breaker.in_flight_requests.trip.total
* es.breaker.eql_sequence.trip.total
* es.breaker.in_model_inference.trip.total

Each of the metrics is exposed at node level.

alex-spies committed:
Rationalize ref-counting around ChannelActionListener (2nd attempt) (#102638)

* Revert "Revert "Rationalize ref-counting around ChannelActionListener (#102551)" (#102609)"

This reverts commit fdf51ac8e92552d68e0e24e465069ec56dc739b4.

* Fix response ownership in ESQL

DaveCTurner committed:
Record chunked response size before sending last chunk (#102586)

We expect that the HTTP response stats should reflect the responses
received by the client, but today we record that a chunked-encoded
response was sent _after_ sending the last chunk, which might mean that
the client can receive the complete response body and then retrieve
stats that do not include that response. With this commit we record the
stats before sending the last chunk.

Closes #102547

original-brownbear committed:
Misc dead code cleanup o.e.e.search (#102617)

Misc obvious cleanup of unused methods and unused parameters.

rockdaboot committed:
[Profiling] Rename request param custom_cost_factor into aws_cost_factor (#102643)

tlrx committed:
Fix InternalEngineTests.testMergeThreadLogging (#102640)

The test has to wait for all merge thread log messages (include Lucene ones) 
to be seen before reset the log level and stop the appender.

Previous attempt wasn't enough: the test failure in #90071 (comment) shows 
that Lucene's merge scheduler thread can log after the Elasticsearch one, and 
if the appender has been closed in the meanwhile it can fail the test.

This change ensure that the appender saw the Merge Scheduler end message 
before resetting the log level and close the engine.

Closes #90071

pmpailis committed:
[CI] Fix for mlt/25_docs_one_shard/Basic mlt query with docs - explicitly on same shard (#102506)

dnhatn committed:
Track blocks when hashing single multi-valued field. (#102612)

DaveCTurner committed:
Mark OwningChannelActionListener for removal (#102647)

Make IntelliJ's warnings about using this workaround more strident.

javanna committed:
Move terms parallel collection tests to TermsTests (#102648)

They were before in AggregatorFactoriesTests, but TermsTests is a more appropriate place for them.

alex-spies committed:
SQL: Prepare integration tests for Serverless testing (#102578)

* Separate tests into those that need frozen indices and those that can do
  without.                                                                                                       
* Make http clients use basic auth if system property
  tests.rest.cluster.multi.user (and ....password) is set.

DaveCTurner committed:
Remove streams from SnapshotInProgressAllocationDecider (#102651)

Replaces the sequence of stream combinators with a simple loop. Also
makes the `THROTTLE` message slightly more informative.

davidkyle committed:
Rest Tests for the mock inference service (#102469)

Adds a qa project to the inference plugin for REST tests
and the mock service plugin TestInferenceServicePlugin

tlrx committed:
Convert range size to int only once (#102397)

Those settings are static, no need to convert them to int in the getters

fcofdez committed:
Allow executing multiple periodic flushes while they are being made durable (#102571)

Closes ES-7200

jan-elastic committed:
Mute ESClientYamlSuiteTestCase (#102661)

abdonpijpelink committed:
[DOCS] Improve enrich policy execute 'wait_for_completion' docs (#102291)

* [DOCS] Improve enrich policy execute 'wait_for_completion' docs

* Update docs/reference/ingest/apis/enrich/execute-enrich-policy.asciidoc

Co-authored-by: James Rodewig <james.rodewig@elastic.co>

---------

Co-authored-by: James Rodewig <james.rodewig@elastic.co>

rockdaboot committed:
[Profiling] Add new request param 'cost_per_core_hour' (#102654)

javanna committed:
Address test issues in TermsTests#testSupportsConcurrentExecution (#102658)

The cardinality function needs to return the same value from both calls, or
it ends up comparing two different scenarios.

Closes #102652

brianseeders committed:
Revert "Mute ESClientYamlSuiteTestCase (#102661)"

This reverts commit e24a909f8bee1166a5317aac9208919d92e40f37.

javanna committed:
Mute failing pit with index filter ccs yaml tests (#102666)

search/350_point_in_time/point-in-time with index filter is failing
often, hence this commit mutes it.

Relates to #102596

thecoop committed:
Inline the tracer into telemetryservice for node construction (#102662)

original-brownbear committed:
Cleanup GetResponse (#102611)

We will want to ref-count this one eventually to save memory.
We can take a few steps towards that goal here today.
Removing the test-only `byte[]` source return that doesn't play well
with pooled + making fields final where possible.

tveasey committed:
[ML] Improve stability of spike and dip detection for the change point aggregation (#102637)

This makes some changes to:
1. Improve the consistency of spike and dip detection
2. Reduce the sensitivity, so that statistically significant spikes and dips are more
    visually obvious in the charts
Previously, we were using cross-validation to set the KDE bandwidth and finding the
exact bandwidth selected by this process could be fairly different for visually similar
time series. This leads to inconsistent detection. In practice, we also want to over
smooth the KDE  (from and ISE perspective) so we detect visually obvious spikes
and dips. Therefore, I migrated to Silverman's rule of thumb, which is optimal for
normal data, but tends to use more smoothing for other data distributions when
compared to cross-validation. It is also significantly more stable to the exact data
values, particularly for small sample sizes.

This change also improves the handling of excluded data. Instead, of excluding an
interval centred on the maximum and minimum value we choose the interval which
includes it with maximum (or minimum) sum. In addition, we now use different data
sets for spike and dip detection and only exclude the other extremum. For example,
there were issues when we happened to exclude the second largest value because
it landed in the excluded region in the vicinity of the minimum. This again caused
visually inconsistent detection.

davidkyle committed:
Add OpenAI error response handler  (#102623)

Creates more specific error messages depending on the cause
and incorporate the OpenAI error message in the response

nik9000 committed:
ESQL: Shrink toString for OutputOperator (#102499)

We sync a status object every thousand iterations or so and if the
`toString` for the operators is super big they take a bunch of time.
Like, the pages have to be small and the `toString` has to be really
expensive for it to matter. Well, `OutputOperator` can have a very big
`toString` because it lists every column it'll write. If there are
thousands of columns this can cost. Well, a little. In a kind of broken
benchmark it's about 3% of the runtime. But shrinking the `toString` is
pretty easy, so let's do it.

original-brownbear committed:
Remove some more `ActionType` subclasses (#102354)

Removing a couple more of these from the search package.

Co-authored-by: Joe Gallo <joegallo@gmail.com>

nik9000 committed:
ESQL: Add newly generated signature

We're generating it now, we should commit it so the docs can have it.

iverase committed:
Remove explicit SearchResponse references from org.elasticsearch.indices.DateMathIndexExpressionsIntegrationIT (#102669)

volodk85 committed:
Line-up snapshot listeners resolution to avoid double calls (#102439)

Snapshot listeners can be concurrently resolved from two different
*clusterApplierService* and *masterService* task threads. If a listener
is a **single action** listener, meaning that it has to be resolved only
once, the following traces can occur, see #101876

```
at org.elasticsearch.action.ActionListener$4.assertFirstRun(ActionListener.java:324)
at org.elasticsearch.action.ActionListener$4.onFailure(ActionListener.java:335)
```

Fix

Line up *resolve listener and remove it from tracking collection*
operations over snapshot listeners in order to avoid double invocation
by separate threads

Fix for #101876

original-brownbear committed:
Fix RefCountedFuture accidentally releasing objects (#102674)

We need to optimistically increment the reference here. Otherwise, we
might release on the thread waiting on the future before we increment on
the future-completing thread, causing a broken ref-count of 0.

>non-issue since this hasn't made it into a release yet.

DaveCTurner committed:
Remove exception-mangling in ESTestCase#safeAwait (#102670)

The static `PlainActionFuture#get` utility actually wraps up a call to
`PlainActionFuture#actionGet` which unwraps exceptions sometimes,
eliding useful information about the source of the problem. This commit
moves to using `Future#get`, with assertions that no exception is thrown
that capture the full stack trace always.

DaveCTurner committed:
Migrate SnapshotResiliencyTests to SubscribableListener (#102668)

There's no need for the exception-mangling behaviour of
`ListenableFuture` in these tests, and if we use `SubscribableListener`
then we'll be able to use various other useful tools in future work.

martijnvg committed:
Minor cleanup of org.elasticsearch.aggregations package (#102622)

* Convert Collections.sort() to List.sort()
* Use Map.computeIfAbsent()
* Use primitive double over Double
* Replace some lambdas with method references.
* Replaces for loops with index variable to use foreach iteration style for loops.

ldematte committed:
Adding extension point in ServerCli to add processing after server process exited (#102475)

ywangd committed:
Include product name in net-new system indices exception (#102678)

This PR includes the product value passed via the header in the
exception message for accessing net-new system indices similar to the
error message for accesing system data streams.

ldematte committed:
Adding option to debug the cli launcher (#102464)

* Adding debug args to cli tools
* Adding instructions

original-brownbear committed:
Remove unused constructors and full classes of RequestBuilders (#102618)

Many of these constructors and a few of the builders themselves have
become unused as a result of recent cleanups. This removes those
constructors and builders.

szabosteve committed:
[DOCS] Fixes link in PUT datafeeds API docs. (#102684)

javanna committed:
Disable parallelism for composite agg against high cardinality fields  (#102644)

We have observed that terms agg against high cardinality fields don't get much gain from parallelism.
On the contrary, they may cause overhead due to work duplication across the different threads.

This commit disables parallelism for composite agg when the source is terms against a high cardinality field, or a script.

jan-elastic committed:
Bugfix: create the tmp dir if it doesn't exist (#102599)

* Make sure the tmp dir exists before creating a tmp file

* Correct file permissions

* changelog

* Recreate tmp dir before creating named pipes

luigidellaquila committed:
Fix DISSECT with empty patterns (#102580)

Fixes https://github.com/elastic/elasticsearch/issues/102577

Ignore empty DISSECT patterns, eg. `DISSECT a "%{a} %{} %{b}"` will only
return two fields (`a` and `b`) and will ignore the empty pattern
(discarding the corresponding values).

breskeby committed:
Make Legacy test cluster infrastructure configuration cache compatible (#101903)

* Move TestCluster plugin and module setup from Node to Cluster
* Do not materialize unused TestCluster
* Fix lazy test cluster evaluation
* Register artifact transforms ones per project
* Make task caching for TestClusterAware tasks CC compatible
* Move stateful logic out of taskgraph.whenready

pgomulka committed:
Update APM agent version (#102691)

this commit upgrades APM agent version to 1.44 to include a fix to a max metric name being increased from 63 to 255

davidkyle committed:
[ML] Returns multiple models for GET _inference (#102639)

original-brownbear committed:
Fix reference count handling in TransportSingleShardAction (#102696)

Just like in the search service, we need to only release the reference after
passing the supplied object to the listener.
Since this is the same pattern in the `TransportSingleShardAction` and in
`SearchService` and we will likely have more spots like this in the near
future I extracted the logic to a new utility.

breskeby committed:
Port S3 qa projects to new test framework (#102555)

This removes the legacy test cluster based plugin usage with our new
testing framework for the searchable snapshots s3 QA project.

- This also cleans up s3-fixture to not rely on docker compose and the test fixture gradle plugin

iverase committed:
Use ElasticsearchAssertions#asserResponse for `MultiSearchResponse` (#102694)

ldematte committed:
REST test framework: add historical features from production code (#102660)

- Check if we can use features from prod code (legacy tests do not support them)
- When OK, TestFeatureService uses ESRestTestCaseHistoricalFeatures as a provider
- Specific error messages for legacy/new tests

williamrandolph committed:
Bring packaging tests README up to date (#102675)

szabosteve committed:
[DOCS] Expands inference API docs (#102634)

* [DOCS] Expands inference API docs.

* [DOCS] Adds example API call.

* [DOCS] Changes task_settings collapsible section.

javanna committed:
Enable randomizations in CompositeAggregationBuilderTests (#102645)

There were two randomizations that have been off for a long time, but there should
be no reason for that. This commit re-enables the randomizations.

jfreden committed:
Make api_key.delete.interval a dynamic setting (#102680)

This PR changes the `xpack.security.authc.api_key.delete.interval`
setting to be a dynamic setting to address
https://github.com/elastic/elasticsearch/issues/92405. 

Relates: #92405

ldematte committed:
Added GC known issue (#102663)

jonathan-buttner committed:
[ML] Inference Service Refactoring results format (#102429)

* Adding results

* Fixing merge issues

* Understanding the complexity

* Making progress on tests

* Tests working

* Some comments

* More comments

* Addressing pr feedback

* Fixing test

* Fixing test

* Fixing up comments and dead code

---------

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

timgrein committed:
Add Connector Sync Job POST API

timgrein committed:
Add Connector Sync Job POST API

elasticsearchmachine commented:
Pinging @elastic/ent-search-eng (Team:Enterprise Search)

timgrein committed:
Remove TODO from ConnectorTestUtils.

timgrein committed:
Improve test method names

timgrein committed:
Remove TODO

nik9000 committed:
ESQL: Share constant null Blocks (#102673)

This speeds up loading empty fields by sharing the `ConstantNullBlock`
across all colunms that are always null. It turns out that creating
these constant null blocks can get expensive - but mostly when you are
creating thousands of them per Page. But that's a thing that happens!

The overhead happens to be related to memory tracking - these are many
small allocations that trigger the real memory breaker a lot. Even if
that weren't true the sharing is pretty simple so let's do it!

This does speed up performance a bit:

```
|                 Min Throughput |   7.23463 |   8.13522 |   0.90058 | ops/s |  +12.45% |
|                Mean Throughput |   9.32052 |  10.7259  |   1.40541 | ops/s |  +15.08% |
|              Median Throughput |   9.52468 |  10.9539  |   1.4292  | ops/s |  +15.01% |
|                 Max Throughput |   9.68817 |  11.3629  |   1.6747  | ops/s |  +17.29% |
|        50th percentile latency |  64.4631  |  51.8855  | -12.5776  |    ms |  -19.51% |
|        90th percentile latency |  74.1094  |  59.6878  | -14.4215  |    ms |  -19.46% |
|        99th percentile latency |  84.0474  |  70.384   | -13.6634  |    ms |  -16.26% |
|      99.9th percentile latency |  98.6236  |  89.8988  |  -8.72481 |    ms |   -8.85% |
|       100th percentile latency | 120.631   | 106.72    | -13.9113  |    ms |  -11.53% |
|   50th percentile service time |  64.4631  |  51.8855  | -12.5776  |    ms |  -19.51% |
|   90th percentile service time |  74.1094  |  59.6878  | -14.4215  |    ms |  -19.46% |
|   99th percentile service time |  84.0474  |  70.384   | -13.6634  |    ms |  -16.26% |
| 99.9th percentile service time |  98.6236  |  89.8988  |  -8.72481 |    ms |   -8.85% |
|  100th percentile service time | 120.631   | 106.72    | -13.9113  |    ms |  -11.53% |
|                     error rate |   0       |   0       |   0       |     % |    0.00% |
```

dnhatn committed:
Fix leaking blocks in TopN (#102715)

Closes #102646

rockdaboot committed:
[Profiling] Remove host metadata cache (#102687)

dnhatn committed:
Add unsupported error message for cross-clusters query (#102677)

Return ES|QL does not yet support querying remote indices for 
cross-clusters queries until the feature is implemented.

Closes #102650

pgomulka committed:
Add apm api for asynchronous counters (always increasing) (#102598)

asynchronous counters (available in otel) are instruments which allow to update the counter's value with the callback style api. The callback has to report always the latest value. Upon restart the value might go back to 0, but in the backend apm server will know that since the value is always increasing.

dnhatn committed:
Fix leaking blocks in BlockUtils (#102716)

javanna committed:
Enable concurrency for multi terms agg  (#102710)

This commit enables concurrency for multi terms agg if all sub-terms agg
target low cardinality fields.

nik9000 committed:
ESQL: Load text field from parent keyword field (#102490)

This adds support for loading a text field from a parent keyword field.
The mapping for that looks like:

```
"properties": {
  "foo": {
    "type": "keyword",
    "fields": {
      "text": { "type": "text" }
    }
  }
}
```

In this case it's safe to load the `text` subfield from the doc values
for the `keyword` field above.

Closes #102473

kderusso commented on x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/400_connector_sync_job_post.yml:
```diff
@@ -0,0 +1,62 @@
+          job_type: full
+          trigger_method: on_demand
+  - set:  { id: id }
+  - match: { id: $id }
```
I hadn't seen this used in practice before, pretty cool! Is the idea to eventually check this with a GET request?

timgrein replied:
Yes, that will be included in the follow-up PRs, when `GET` is implemented.

kderusso commented on x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/400_connector_sync_job_post.yml:
```diff
@@ -0,0 +1,62 @@
+      features: warnings
+  - do:
+      warnings:
+        - Missing field 'job_type'. Using default value 'full'.
```
Is it standard practice to return header warnings when we use default values? Would documentation suffice here?

timgrein replied:
I don't know, if it's standard practice, but I've seen it in other places. I thought it could be helpful, so you're aware of values set in the background, which could introduce side effects you don't have in mind. WDYT?

timgrein replied:
Ok you both sold it to me :) Adjusted it with https://github.com/elastic/elasticsearch/pull/102714/commits/46dccc925b277bcbfe8a141c6d704d4e6bf7c8ef.

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearch.java:
```diff
@@ -105,6 +108,8 @@ public class EnterpriseSearch extends Plugin implements ActionPlugin, SystemInde
 
     public static final String CONNECTOR_API_ENDPOINT = "_connector";
 
+    public static final String CONNECTOR_SYNC_JOB_API_ENDPOINT = CONNECTOR_API_ENDPOINT + "/_sync_job";
```
Is this the right place for this endpoint name? Should it be in the post command itself? Typically we have higher level definitions here?

timgrein replied:
Fair point, adjusted it with https://github.com/elastic/elasticsearch/pull/102714/commits/91da1f343d1851876fb10a7b475d0a1b09841c09.

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJob.java:
```diff
@@ -0,0 +1,498 @@
+import java.util.Objects;
+
+/**
+ * Represents a sync job in the Elasticsearch ecosystem. Sync jobs refer to a unit of work, which syncs data from a 3rd party
```
Nice class comment!

timgrein replied:
Thanks! :)

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJob.java:
```diff
@@ -0,0 +1,498 @@
+
+    private final String id;
+
+    private final Long indexedDocumentCount;
```
Any reason these can't be primitives that start at 0?

timgrein replied:
Nope, not really. Adjusted it with https://github.com/elastic/elasticsearch/pull/102714/commits/56e5ce8bbdb1621fd4b0fa90195e2fa823b1acee. Thanks for the hint!

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJob.java:
```diff
@@ -0,0 +1,498 @@
+    @Nullable
+    private final Instant lastSeen;
+
+    private final Map<String, Object> metadata;
```
I'm guessing this might be a good use case for a general Map? Any thoughts on whether it should be an object?

timgrein replied:
From a Kibana type definition perspective `metadata` looks like this `metadata: Record<string, unknown>;
`, so it kinda made sense to me to use `String` as key type and `Object` as value type. 

> I'm guessing this might be a good use case for a general Map?

WDYM with "general Map"? `Map` or `Map<?, ?>`?

> Any thoughts on whether it should be an object?

I'm not 100% sure, if I understood the question correctly. Do you mean to create a separate class `Metadata`?  If so `metadata` doesn't tell us anything about the fields used inside of it, but is a generic field, which could contain any key-value pairs.

kderusso replied:
I think this is OK, to keep a `Map<String,Object>` for this use case.

kderusso replied:
Thank you for explaining!

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/action/PostConnectorSyncJobAction.java:
```diff
@@ -0,0 +1,195 @@
+
+    public static final PostConnectorSyncJobAction INSTANCE = new PostConnectorSyncJobAction();
+
+    public static final String NAME = "cluster:admin/xpack/connector/sync_job/post";
```
Why is this `cluster:admin/xpack/connector/sync_job/post` and not something like `cluster:admin/xpack/connector/post_sync_job`?

timgrein replied:
I've checked how the constants are named inside [Constants.java](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/qa/operator-privileges-tests/src/javaRestTest/java/org/elasticsearch/xpack/security/operator/Constants.java) and I found many instances, which put the `post` (or the action in general) after the resource (in this case `sync_job`):

See [this part in Constants](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/qa/operator-privileges-tests/src/javaRestTest/java/org/elasticsearch/xpack/security/operator/Constants.java#L26-L42) for example. But I don't hold a strong opinion here. Do we've a best practice for the ent search action names?

kderusso replied:
OK that's fair, thanks.

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/action/RestPostConnectorSyncJobAction.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Override
+    public List<Route> routes() {
+        return List.of(new Route(POST, "/" + EnterpriseSearch.CONNECTOR_SYNC_JOB_API_ENDPOINT));
```
I think we want to do some refactoring here, see comments in EnterpriseSearch

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/action/RestPostConnectorSyncJobAction.java:
```diff
@@ -0,0 +1,47 @@
+        return channel -> client.execute(
+            PostConnectorSyncJobAction.INSTANCE,
+            request,
+            new RestToXContentListener<>(channel, r -> RestStatus.CREATED, r -> null)
```
Will this always be a 201, or can it sometimes be a 200?

timgrein replied:
I think this returns either `201` or it fails with some `4XX` status as this endpoint is only used for creating and not updating sync jobs. Citing this [StackOverflow post](https://stackoverflow.com/a/44396004) and this [rfc](https://datatracker.ietf.org/doc/html/rfc7231#section-6.3.2) it seems to me that `201` indicating that something was created + returning a reference to the created object (in this case `id`) seems to favour `201` over `200`, when the request was successful.

WDYT?

kderusso replied:
I'm good with that.

kderusso reviewed (commented):
First pass review, looking really good!

brianseeders committed:
Set DEBIAN_FRONTEND=noninteractive before apt-get commands

brianseeders committed:
Use NEEDRESTART_MODE instead of DEBIAN_FRONTEND

brianseeders committed:
Move NEEDRESTART_MODE

idegtiarenko committed:
Simplify FailedShardEntry formatting (#102681)

Avoid creating temporary list when converting FailedShardEntry to string.
Also simplify other format usage in the surrounding class.

timgrein committed:
Move CONNECTOR_SYNC_JOB_API_ENDPOINT from EnterpriseSearch to PostConnectorSyncJobAction.

n1v0lg committed:
Fix trial license version check (#102362)

This PR fixes the version check we perform to determine if a new trial
can be started. 

For trial licenses, https://github.com/elastic/elasticsearch/pull/100169
decoupled version tracking from the Elasticsearch version. In short, the
new logic is to allow a new trial license each time we increment the
trial license version counter (with some additional caveats around BWC).
This means we should allow a new trial license if the version the
previous trial was started on is _below_ the `CURRENT` version. This
condition was previously reversed. This PR fixes the check.

I've also opted to restrict `ableToStartNewTrial` to not take arbitrary
input and instead to always compare to `CURRENT`. This allows us to
avoid the unnecessary complexity of handling a BWC edge case where both
trial versions are below the cut-over value (this is impossible in
practice).  

Fixes: https://github.com/elastic/elasticsearch/issues/102286

timgrein committed:
Use primitive type longs for counts

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java:
```diff
@@ -0,0 +1,169 @@
+                    try {
+                        String syncJobId = generateId();
+
+                        final IndexRequest indexRequest = new IndexRequest(CONNECTOR_SYNC_JOB_INDEX_NAME).id(syncJobId)
```
```suggestion
                        final IndexRequest indexRequest = new IndexRequest(CONNECTOR_SYNC_JOB_INDEX_NAME).id(UUIDs.base64UUID())
```

this is what generating the ID does under the hood, I think it's good to use it here

timgrein replied:
Adjusted with https://github.com/elastic/elasticsearch/pull/102714/commits/c21492cece4fa8e94c061f360e75fc0552cebb3a. I'm still keeping it inside the method as it explains the reasoning, why we do this "low-level" call by ourselves.

carlosdelest reviewed (commented):
Very good job, Tim! 👏 

I left some minor comments. I think we should remove the warning headers for default values, the rest LGTM.

jedrazb commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobType.java:
```diff
@@ -0,0 +1,31 @@
+            }
+        }
+
+        return null;
```
Should we fail fast if we are provided illegal `syncJobTypeString`?  Should we throw `IllegalArgumentException` if we are unable to parse the string?

jedrazb commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobTriggerMethod.java:
```diff
@@ -0,0 +1,30 @@
+            }
+        }
+
+        return null;
```
Should we fail fast if we are provided illegal `triggerMethodString`?  Should we throw `IllegalArgumentException` if we are unable to parse the string?

timgrein replied:
That makes sense! Adjusted with https://github.com/elastic/elasticsearch/pull/102714/commits/3536d74ea02921008d016b029ac176e7e54d978c.

pgomulka committed:
Fix wrong metric type in a test (#102729)

timgrein committed:
Remove header warnings on missing trigger_method and job_type.

timgrein commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -172,22 +175,22 @@ public Connector(StreamInput in) throws IOException {
+    public static final ParseField PIPELINE_FIELD = new ParseField("pipeline");
     private static final ParseField SCHEDULING_FIELD = new ParseField("scheduling");
-    private static final ParseField SERVICE_TYPE_FIELD = new ParseField("service_type");
+    public static final ParseField SERVICE_TYPE_FIELD = new ParseField("service_type");
```
I wouldn't say so no. As the `ConnectorSyncJob` simply references a subset of fields from `Connector` I don't see a good intermediate class, which would make sense to me at least.

carlosdelest replied:
Makes sense. I would probably copy the fields over to the classes as there is no direct dependency on the class then.

timgrein replied:
The thing is that I'm using those fields also in `ConnectorSyncJobIndexService`, when deserialising those fields. If they change inside `Connector` someone needs to do changes in two places (`ConnectorSyncJob` and `Connector`) or run into errors first. WDYT?

kderusso replied:
I wonder if refactoring these into something like a `ConnectorSharedFieldDefinition` would be how to approach this?

timgrein commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJob.java:
```diff
@@ -0,0 +1,498 @@
+ */
+public class ConnectorSyncJob implements Writeable, ToXContentObject {
+
+    public static final ParseField CANCELATION_REQUESTED_AT_FIELD = new ParseField("cancelation_requested_at");
```
I'm referencing them somewhere else (tests) to check values, so I guess it's fine to have them public?

carlosdelest replied:
If they're being referenced by tests, making them package private would make more sense to me (no access modifier)

timgrein replied:
Good point, adjusted with https://github.com/elastic/elasticsearch/pull/102714/commits/ec40aebdbe1a4f0104648d444983eb4bf03e819a.

timgrein commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java:
```diff
@@ -0,0 +1,169 @@
+                public void onResponse(Connector connector) {
+                    Instant now = Instant.now();
+                    ConnectorSyncJobType jobType = Objects.requireNonNullElse(request.getJobType(), ConnectorSyncJob.DEFAULT_JOB_TYPE);
+                    ConnectorSyncJobTriggerMethod triggerMethod = Objects.requireNonNullElse(
```
IMHO it's a bit better to have them here. As we've two constructors inside the request we need to keep two places in sync instead of one in the index service.

timgrein committed:
Update x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/action/PostConnectorSyncJobAction.java

Co-authored-by: Carlos Delgado <6339205+carlosdelest@users.noreply.github.com>

timgrein committed:
Update x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/action/TransportPostConnectorSyncJobAction.java

Co-authored-by: Carlos Delgado <6339205+carlosdelest@users.noreply.github.com>

timgrein committed:
Use parse field constants in PostConnectorSyncJobAction

timgrein commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/action/PostConnectorSyncJobAction.java:
```diff
@@ -0,0 +1,195 @@
+        );
+
+        static {
+            PARSER.declareString(constructorArg(), new ParseField("id"));
```
Adjusted with https://github.com/elastic/elasticsearch/pull/102714/commits/7c4e94710b621514a04f34ae6b1b0a56f861b5d5.

timgrein committed:
Use listener.delegateFailure

timgrein commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java:
```diff
@@ -0,0 +1,169 @@
+        ActionListener<PostConnectorSyncJobAction.Response> listener
+    ) {
+        try {
+            getSyncJobConnectorInfo(request.getId(), new ActionListener<>() {
```
Good point! Adjusted with https://github.com/elastic/elasticsearch/pull/102714/commits/3e2373c926db03aeefa317f7f722bd2ac48cf69b.

timgrein committed:
Generate a UUID directly

timgrein committed:
Remove unnecessary constant.

timgrein commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexServiceTests.java:
```diff
@@ -0,0 +1,221 @@
+
+public class ConnectorSyncJobIndexServiceTests extends ESSingleNodeTestCase {
+
+    private static final Integer ZERO = 0;
```
Adjusted with https://github.com/elastic/elasticsearch/pull/102714/commits/c7bb1420300634b6c6c93934df0a16f152a575cc.

timgrein committed:
Fix formatting

thecoop committed:
Pull out discovery module construction method in node construction (#102569)

timgrein committed:
Make fields package private

pgomulka committed:
Reuse running mock apm server (#102698)

When a test cluster in serverless is run with --with-apm-server
3 nodes were attempting to start apm server. This was throwing an exception.

There is no harm in reusing the already running apm server,
and in fact is more realistic

timgrein committed:
Require non-null value for "id" inside ConnectorSyncJob.

timgrein commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJob.java:
```diff
@@ -0,0 +1,498 @@
+        this.deletedDocumentCount = deletedDocumentCount;
+        this.error = error;
+        this.id = id;
+        this.indexedDocumentCount = indexedDocumentCount;
```
As the counts are now primitive types, they shouldn't be `null`, but `0` I suppose, which is fine. `error` is already `nullable`. `id` should never be null, that's right! (addressed with https://github.com/elastic/elasticsearch/pull/102714/commits/d2a5f385a4600e2d1753b974a6105db021ee87d9).

danielmitterdorfer committed:
[Profiling] Report in status API if docs exist (#102735)

With this commit we add a new field `has_data` to the profiling status
API. This field is true iff there is at least one profiling event. As
profiling events are the aggregate root in the profiling domain model,
it is sufficient to check for presence of documents in that data stream.

martijnvg committed:
Introduce fielddata cache ttl (#102682)

Add indices.fielddata.cache.expire setting to control how long field data entries can stay in the cache after last access.
And enabling cache expiring after last access by default. After 1 hour of last access.

The reasoning of this change is that global ordinals can sometimes take non-trivial amount of memory.
And sometimes searches hit older data that isn't accessed later anymore. However this does keep cached global ordinal around, which can take non-trivial amount of memory. The idea is that if global ordinals isn't used after 60 minutes of last access it is allowed to be purged.

kkrik-es committed:
Use default_shards in downsampling (#102743)

Index patter in a TSDS conflicts with a global template matching all
indexes, `(global => [*])`. This is sometimes added in
ESClientYamlSuiteTestCase, leading to flaky warnings. To avoid this, use
feature default_shards.

Fixes #102489

timgrein committed:
Throw IllegalArgumentException on unknown trigger method

timgrein committed:
Reuse already existing methods

timgrein committed:
Fix actual and expected order

timgrein committed:
Throw IllegalArgumentException on invalid connector sync job type string.

timgrein commented:
Thanks @kderusso @carlosdelest and @jedrazb for the detailed review, already learned a lot around ES development! I think I've addressed all of your comments. Let me know, if there's anything else, which needs to be addressed!

timgrein committed:
Fix formatting

amyjtechwriter committed:
Adding image to 'basic write model' section. (#102702)

jedrazb reviewed (approved):
To me it looks good! Given that:
- it's under a feature flag
- e2e tests show that it's working
Let's ship 🚀

iverase committed:
Disable concurrency when synthetic source is enabled (#102748)

timgrein committed:
Handle null values on the request level as sync job or trigger method not specified.

timgrein committed:
Remove TODO, which was already addressed.

timgrein committed:
Add integration tests for creating a sync job once with an invalid job type and once with an invalid trigger method.

benwtrent committed:
Adding assertion and message to debug issue #102261 (#102733)

timgrein committed:
Throw IllegalArgumentException on invalid sync status string and also add tests.

eyalkoren committed:
Switch default logs template to search all fields by default (#102456)

jan-elastic committed:
Improve Processors' exception message. (#102760)

rockdaboot committed:
[Profiling] Add per_core_watt_amd64 request param (#102738)

* [Profiling] Add per_core_watt_amd64 request param

* Test perCoreWattARM64 serialization

* Rename per_core_watt to per_core_watt_x86

* Improve request serialization test

thecoop committed:
Symlinks on windows needs admin privileges (#102762)

so disable the test when it throws an exception due to missing privileges

jan-elastic committed:
Treat zero ML processors correctly (#102742)

I'll postpone merging this until there's a verdict on #102739.

bpintea committed:
ESQL: remove `time_zone` request parameter (#102767)

This removes the parameter `time_zone` from the request. It's kept in
the configuration and always initialized to `UTC`.

The parameter will be added back later with broader support for time
zones.

nik9000 committed:
ESQL: Load stored fields sequentially (#102727)

The lucene APIs can load stored fields more quickly if flip them into
"sequential" mode - but it's only faster when you are loading a dense
set of documents - all bunched up. This flips the stored field loading
code to use a sequential reader when the documents are fully dense -
literally just counting without gaps - which is precisely what the fetch
phase does.

kderusso commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexServiceTests.java:
```diff
@@ -0,0 +1,220 @@
+        ActionFuture<DocWriteResponse> index = client().index(indexRequest);
+
+        // wait 10 seconds for connector creation
+        index.get(TIMEOUT_SECONDS, TimeUnit.SECONDS);
```
To guard against future potential flakey tests, will this ensure that a timeout fails with the appropriate message?

kderusso commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexServiceTests.java:
```diff
@@ -0,0 +1,220 @@
+            }
+        });
+
+        if (exception.get() != null) {
```
We may want to log whether there was a timeout first, before we throw.

original-brownbear committed:
Add leak-tracked ref counting to MultiSearchResponse (#102479)

Make the ref-counting leak tracked and release child items for
MultiSearchResponse and adjust tests + REST action accordingly.

kderusso commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobTestUtils.java:
```diff
@@ -0,0 +1,88 @@
+
+    public static ConnectorSyncJob getRandomConnectorSyncJob() {
+        Instant lowerBoundInstant = Instant.ofEpochSecond(0L);
+        Instant upperBoundInstant = Instant.ofEpochSecond(3000000000L);
```
We're good until the year 2065 😉

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java:
```diff
@@ -0,0 +1,159 @@
+        }
+    }
+
+    private String generateId() {
```
(Assuming the answer to your TODO is "no it's not needed") It looks like `DocWriteResponse` has a `getId()` method, could we use that for the response and get rid of this?

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java:
```diff
@@ -0,0 +1,159 @@
+    private String generateId() {
+        /* Workaround: only needed for generating an id upfront, autoGenerateId() has a side effect generating a timestamp,
+         * which would raise an error on the response layer later ("autoGeneratedTimestamp should not be set externally").
+         * TODO: do we even need to copy the "_id" and set it as "id"?
```
It depends on what you'll be using these IDs for. If you're only looking up by ID, I would think the ID would be sufficient. If you want to do things like aggregate by ID, you need to store it as a variable.

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java:
```diff
@@ -0,0 +1,159 @@
+     * @param request   Request for creating a connector sync job.
+     * @param listener  Listener to respond to a successful response or an error.
+     */
+    public void createConnectorSyncJob(
```
I feel like this method is running into a trap that we ran into with search applications and again with synonyms, but never clearly resolved - potential race conditions if some items are added/updated/deleted midway through this call. This is "OK" (for now) for the management CRUD APIs because they're expected to not be used that frequently, but I worry that this pattern could be problematic here with sync jobs as presumably they're a lot more frequent. @carlosdelest - I'd like to hear your thoughts here as well?

kderusso replied:
To be clear: Since this is tech preview I think it could be OK to commit now and fix later.

carlosdelest replied:
Agreed - but we don't have a clear pattern for getting out of this situation. I'd say we're good with this for now.

kderusso reviewed (approved):
I have some questions and concerns, but nothing that can't be dealt with in a followup PR. Thank you for iterating!

dnhatn committed:
Allow match field in enrich fields (#102734)

It's perfectly fine for the match field of an enrich policy to be 
included in the enrich fields. However ESQL enrich consistently fails on
such an enrich policy because it mistakenly excludes the match field
from the enrich mapping during resolution.

DaveCTurner committed:
Use random purpose in blob store repository tests (#102789)

Today many blob store repository tests specify that the operations they
perform have purpose `OperationPurpose#SNAPSHOT`, but most of these
tests do not care about the purpose of these operations. This commit
switches them to using a random purpose to highlight that the purpose is
unimportant to the test.

danielmitterdorfer committed:
[Profiling] Notify early about task cancellation (#102740)

With this commit we add explicit checks in selected places of
`TransportGetStackTracesAction` whether the parent task has been
cancelled. This can happen for example when a request times out and the
client closes the TCP connection. While parent task cancellation is
checked explicitly before making internal calls (such as an mget) this
can lead to excessive log output due to the nature of these requests (we
usually send multiple of them concurrently). With the explicit checks,
the listener is notified as early as possible and request processing is
cancelled, thus reducing log output. Note that this is only best effort
as a task might be cancelled in between our explicit check and the one
done by the internal client.

ywangd committed:
Record S3 retry stats with APM metrics (#102505)

S3 retry stats are added in #100272 using log messages. Changes in this
PR are intended to replace the log based stats with APM metrics.

jedrazb committed:
[Connector API] Implementation of list, get, delete connector endpoints (#102592)

Add list, get, delete endpoints and parsers for Connector document data structures, ensuring payload validation and parsing. Includes BWC, Yaml, unit tests for verification.

carlosdelest reviewed (approved):
Good job @timgrein ! 💯

original-brownbear committed:
Add simple stress test to EnrichMultiNodeIT (#102751)

Add a simple test that forces enrich tasks to queue up to guard against
bugs around holding on to multi search respones in this logic, now that we
are introducing ref counting to them.

relates #102479

timgrein committed:
Make some fields public again

timgrein commented:
Reopen
